### PR TITLE
Add logging checks

### PR DESCRIPTION
**Pull request by wshearn**
The ocutil changes are so I can use oc exec and be able to easily parse the output. As seen on line 76 in cron-send-logging-checks.py

### CODE REVIEW

**Comment by wshearn** [comment]
@kwoodson when you get a chance Monday can you review this?

File: scripts/monitoring/cron-send-logging-checks.py
Comment (portante) [inline scripts/monitoring/cron-send-logging-checks.py]: Could consider using `try/except/else` here and placing the `print res` under the else clause.

File: scripts/monitoring/cron-send-logging-checks.py
Comment (portante) [inline scripts/monitoring/cron-send-logging-checks.py]: By the time this loop is done, we appear to have iterated the nodes twice, and the pods twice, and then iterated the fluentd pods for each node again.

Perhaps we can consider a way to avoid the number of iterations over the data.

File: scripts/monitoring/cron-send-logging-checks.py
Comment (portante) [inline scripts/monitoring/cron-send-logging-checks.py:106]: Looks like you can exit the loop once you found a pod with a matching name.

**Review by portante — COMMENTED** [review]
Looks good, fluentd check seems like it might not scale with high node/pod counts.

File: scripts/monitoring/cron-send-logging-checks.py
Comment (wshearn) [reply scripts/monitoring/cron-send-logging-checks.py]: Eh, we only care if the site returns a <500 status code. Though the print shouldn't have even slipped in.

**Review by portante — COMMENTED** [review]
Looks good!

File: scripts/monitoring/cron-send-logging-checks.py
Comment (twiest) [inline scripts/monitoring/cron-send-logging-checks.py]: I worry about the reliability of doing remote execution (oc exec) for health checks. AFAIK, this is the first time we've considered using it for health checks.

I feel the same way about remote execution based health checks as I do about remote autoheals. Basically, I think we should do them locally unless we absolutely can't.

Here are a few specifics:
- oc exec runs through the masters. If the masters are down, it's going to alert saying these are down as well.
- master load causing sporadic failures
- blue/green update causes sporadic failures (even though we should be going through the ELB, which should bleed connections, etc, it doesn't mean it'll be clean).

Is there a reason you can't do these checks from a sidecar container? I believe the card asked for a sidecar container as well.

What were the decision points on making this a remote health check?

File: ansible/roles/os_zabbix/vars/template_openshift_logging.yml
Comment (twiest) [inline ansible/roles/os_zabbix/vars/template_openshift_logging.yml:33]: We need to also monitor the disk usage of the elastic search PV. This, IMO, is the most important thing to monitor as it's the thing I'm sure we'll run out of, possibly quickly.

File: scripts/monitoring/cron-send-logging-checks.py
Comment (twiest) [inline scripts/monitoring/cron-send-logging-checks.py]: What happens if the oc exec succeeds, but the curl itself fails?

Like if elastic search is in a bad state or down for whatever reason?

From my cursory view, that seems like it would cause an ops-runner alert (as this script would back trace).  ops-runner should only alert if there's a bug in the script. The script itself should report typical failure cases.

**Review by twiest — CHANGES_REQUESTED** [review]

File: scripts/monitoring/cron-send-logging-checks.py
Comment (wshearn) [reply scripts/monitoring/cron-send-logging-checks.py]: Even if we did the sidecar container it would still need to query the master to get the ES endpoints as we are running this across all the pods to make sure we are not in a split brain scenario.

As for why I went this way, it was the quickest way to get it up and running, and right now from when I need this done to when I could get the sidecar done there is not enough time.

File: ansible/roles/os_zabbix/vars/template_openshift_logging.yml
Comment (wshearn) [reply ansible/roles/os_zabbix/vars/template_openshift_logging.yml:33]: I knew I forgot one. One PV check coming right up. (also how would checking the diskspace for the pods work with a sidecar container?)

File: scripts/monitoring/cron-send-logging-checks.py
Comment (wshearn) [reply scripts/monitoring/cron-send-logging-checks.py]: Good thinking, I will add a try/except around it and fail the check if it hits that.

File: ansible/roles/os_zabbix/vars/template_openshift_logging.yml
Comment (twiest) [reply ansible/roles/os_zabbix/vars/template_openshift_logging.yml:33]: Since it's a PV, I believe PVs are acessible by all of the POD. I'm not 100% on that, though.

File: scripts/monitoring/cron-send-logging-checks.py
Comment (twiest) [reply scripts/monitoring/cron-send-logging-checks.py]: Doesn't elastic search know when it's in a split brain mode?

Each node in the elastic search cluster would get it's own sidecar container, so each would be monitored.

@a13m What are your thoughts on using remote `oc exec` calls for monitoring checks?

Does this worry you from a check reliability point of view?

File: scripts/monitoring/cron-send-logging-checks.py
Comment (twiest) [inline scripts/monitoring/cron-send-logging-checks.py]: Is this port not exposed on the POD IP?

In other words, could we do this curl directly from the master to the pod?

**Review by twiest — DISMISSED** [review]

File: scripts/monitoring/cron-send-logging-checks.py
Comment (twiest) [reply scripts/monitoring/cron-send-logging-checks.py]: If we can do the curl from the master to the pod IP, then I'd also suggest changing this to pycurl (already inside of the host monitoring container), so that we have more control over the call as well as better error handling / messages.

Comment (wshearn) [reply scripts/monitoring/cron-send-logging-checks.py]: It is, but ssl certs are stored in a secret so it would still depend on the master being up.

File: scripts/monitoring/cron-send-logging-checks.py
Comment (wshearn) [reply scripts/monitoring/cron-send-logging-checks.py]: > Doesn't elastic search know when it's in a split brain mode?
> Sadly, not that I know of. When we ran into it with testing the only way we saw it was the nodes did not match across all ES instances.

**Comment by a13m** [comment]
I certainly understand @twiest's concerns on this, but I think that OpenShift is designed for this method to work, and in the interest of time we should move forward with this.  I do agree that this may not be the right way to go in the long term, and we should watch it closely.

**Comment by twiest** [comment]
@wshearn with @a13m's comments, I'll dismiss that review and start a new one.

File: ansible/roles/os_zabbix/vars/template_openshift_logging.yml
Comment (twiest) [reply ansible/roles/os_zabbix/vars/template_openshift_logging.yml:33]: I think we need the disk check before merging this PR.

@wshearn do you disagree with this?

File: scripts/monitoring/cron-send-logging-checks.py
Comment (twiest) [reply scripts/monitoring/cron-send-logging-checks.py]: Understood, it would:
- get us away from doing an oc exec (which goes through the api server and might be flaky), at least for these types of calls
- would allow us more control over the check in general (able to use pycurl, able to check return codes, retries, etc)
- would allow for this to be refactored easier into a sidecar pod (obviously a future version where the secret containing the ssl certs would get mounted in by openshift).
- doesn't require curl to be installed on the pod

**Review by twiest — CHANGES_REQUESTED** [review]

File: scripts/monitoring/cron-send-logging-checks.py
Comment (wshearn) [inline scripts/monitoring/cron-send-logging-checks.py]: @twiest any reason our bot does not like filter?

Comment (twiest) [reply scripts/monitoring/cron-send-logging-checks.py]: This is a pylint check. Here's more detail:
"Long story short, going forward you should prefer List comprehension instead of filter. pylint enforces this with this error message. – thefourtheye Feb 16 '14 at 3:08"

From: http://stackoverflow.com/questions/21806765/how-is-using-filter-a-bad-builtin

File: ansible/roles/os_zabbix/tasks/main.yml
Comment (twiest) [inline ansible/roles/os_zabbix/tasks/main.yml]: I think you mean `openshift_logging` for the tag.

File: scripts/monitoring/cron-send-logging-checks.py
Comment (twiest) [inline scripts/monitoring/cron-send-logging-checks.py]: Isn't this part of the container name generated?  `dy48r5sl`

So, will that work in other clusters?

Will it even work in this cluster if you re-deploy the container?

For the curl above, you used `pod_name`. I'd expect something similar here.

File: scripts/monitoring/cron-send-logging-checks.py
Comment (twiest) [inline scripts/monitoring/cron-send-logging-checks.py]: I'd prefer not to disable these. Instead, I'd prefer the code be refactored a little into smaller functions.

I understand what you're saying in the comment, but these specific pylint checks should never be disabled IMO.

Also, I don't know when we'll get to the sidecar container, so we'll likely be living with this code for some time.

File: scripts/monitoring/cron-send-logging-checks.py
Comment (twiest) [inline scripts/monitoring/cron-send-logging-checks.py]: Same comment as above. I'd prefer these not be disabled.

Also, whenever applying pylint disables, you need to justify why.

**Review by twiest — CHANGES_REQUESTED** [review]

**Comment by twiest** [comment]
:+1:

**Review by twiest — APPROVED** [review]

**Pull request merged**
